File: PokeLangPython/Lexer/Reservada.py
from AFD import AFD
from Token import Token

class Reservada(AFD):
    reservadas = ["eevee", "espeon", "forretress", "poliwhirl", "porygon", "squirtle", "unown", "pokeball", "pokedex", "ditto"]

    def evaluate(self, code, linha, column):
        reservada = ""

        for palavra_reservada in self.reservadas:
            iterator = iter(code)
            i = 0
            try:
                while i < len(palavra_reservada) and next(iterator) == palavra_reservada[i]:
                    reservada += palavra_reservada[i]
                    i += 1
            except StopIteration:
                pass

            if i == len(palavra_reservada) and (not iterator or next(iterator) in (' ', '(', ')', '')):
                return Token("RESERVADA_" + palavra_reservada.upper(), reservada, linha, column)

        return None

File: PokeLangJava/src/Lexer/Reservada.java
package Lexer;

import java.text.CharacterIterator;

public class Reservada extends AFD {
	String reservadas[] = { "eevee", "espeon", "forretress", "poliwhirl", "porygon", "squirtle", "unown", "pokeball", "pokedex", "ditto" };

	@Override
	public Token evaluate(CharacterIterator code, int linha, int column) {
		String reservada = "";

		for (int i = 0; i < reservadas.length; i++) {
			for (char c : reservadas[i].toCharArray()) {
				if (code.current() == c) {
					reservada += code.current();

					code.next();
				}
			}

			if (endReservada(code)) {
				if (reservada.equals(reservadas[i])) {
					return new Token("RESERVADA_" + reservadas[i].toUpperCase(), reservada, linha, column);
				}
			}
		}

		return null;
	}

	private boolean endReservada(CharacterIterator code) {
		return code.current() == ' '
				|| code.current() == '('
				|| code.current() == ')'
				|| code.current() == CharacterIterator.DONE;
	}
}
